lib/handlers/quote/provider-migration/v3/traffic-switch-on-chain-quote-provider.ts

```diff
@@ -0,0 +1,102 @@
+import { IOnChainQuoteProvider, metric, MetricLoggerUnit, RouteWithQuotes } from '@uniswap/smart-order-router'
+import { MixedRoute, V2Route, V3Route } from '@uniswap/smart-order-router/build/main/routers'
+import { Currency, CurrencyAmount } from '@uniswap/sdk-core'
+import { ProviderConfig } from '@uniswap/smart-order-router/build/main/providers/provider'
+import { BigNumber } from 'ethers'
+import { QUOTE_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION } from '../../util/quote-provider-traffic-switch-configuration'
+
+export type TrafficSwitchOnChainQuoteProviderProps = {
+  currentQuoteProvider: IOnChainQuoteProvider
+  targetQuoteProvider: IOnChainQuoteProvider
+}
+
+export class TrafficSwitchOnChainQuoteProvider implements IOnChainQuoteProvider {
+  private readonly currentQuoteProvider: IOnChainQuoteProvider
+  private readonly targetQuoteProvider: IOnChainQuoteProvider
+
+  protected readonly SHOULD_SWITCH_EXACT_IN_TRAFFIC = () =>
+    QUOTE_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.switchExactInPercentage > this.getRandomPercentage()
+  protected readonly SHOULD_SAMPLE_EXACT_IN_TRAFFIC = () =>
+    QUOTE_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.samplingExactInPercentage > this.getRandomPercentage()
+  protected readonly SHOULD_SWITCH_EXACT_OUT_TRAFFIC = () =>
+    QUOTE_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.switchExactOutPercentage > this.getRandomPercentage()
+  protected readonly SHOULD_SAMPLE_EXACT_OUT_TRAFFIC = () =>
+    QUOTE_PROVIDER_TRAFFIC_SWITCH_CONFIGURATION.samplingExactOutPercentage > this.getRandomPercentage()
+
+  constructor(props: TrafficSwitchOnChainQuoteProviderProps) {
+    this.currentQuoteProvider = props.currentQuoteProvider
+    this.targetQuoteProvider = props.targetQuoteProvider
+  }
+
+  async getQuotesManyExactIn<TRoute extends V3Route | V2Route | MixedRoute>(
+    amountIns: CurrencyAmount<Currency>[],
+    routes: TRoute[],
+    providerConfig?: ProviderConfig
+  ): Promise<{
+    routesWithQuotes: RouteWithQuotes<TRoute>[]
+    blockNumber: BigNumber
+  }> {
+    const sampleTraffic = this.SHOULD_SAMPLE_EXACT_IN_TRAFFIC()
+    const switchTraffic = this.SHOULD_SWITCH_EXACT_IN_TRAFFIC()
+
+    let currentQuote
+    let targetQuote
+
+    metric.putMetric('ON_CHAIN_QUOTE_PROVIDER_EXACT_IN_TRAFFIC_TOTAL', 1, MetricLoggerUnit.None)
+
+    if (sampleTraffic) {
+      metric.putMetric('ON_CHAIN_QUOTE_PROVIDER_EXACT_IN_TRAFFIC_SAMPLING', 1, MetricLoggerUnit.None)
+
+      currentQuote = await this.currentQuoteProvider.getQuotesManyExactIn(amountIns, routes, providerConfig)
+      targetQuote = await this.targetQuoteProvider.getQuotesManyExactIn(amountIns, routes, providerConfig)
```

Comment:
We should send these 2 in parallel right?

Reply:
Do we want to collect some latency metrics?

Reply:
I parallelized those two calls. There's also a nuance now, if the target quoter calls throws runtime exception during sampling, we will want to swallow the exception, because otherwise sampling traffic is no longer shadow traffic.